runner/bsp/worker/src/mill/bsp/worker/MillBuildServer.scala

```diff
@@ -60,25 +61,108 @@ private class MillBuildServer(
   protected var sessionInfo: SessionInfo = scala.compiletime.uninitialized
   // Set when the `MillBuildBootstrap` completes and the evaluators are available
   private var bspEvaluators: Promise[BspEvaluators] = Promise[BspEvaluators]()
+  private def bspEvaluatorsOpt(): Option[BspEvaluators] =
+    bspEvaluators.future.value.flatMap(_.toOption)
+  private var savedPreviousEvaluators = Option.empty[BspEvaluators]
   // Set when a session is completed, either due to reload or shutdown
   private[worker] var sessionResult: Promise[BspServerResult] = Promise()
 
   private val requestCount = new AtomicInteger
 
   def initialized = sessionInfo != null
 
-  def updateEvaluator(evaluatorsOpt: Option[Seq[EvaluatorApi]]): Unit = {
-    baseLogger.debug(s"Updating Evaluator: $evaluatorsOpt")
+  def updateEvaluator(
```

Comment:
What's all this new code about? It's not clear to me from reading it

Reply:
Just refactored it a bit, and added comments